#25733 close channels from SelectionHandler to make sure they are properly flushed from Selector

jrudolph opened the pull request:
This makes sure that we report connection closure only after it has really happened. This is important on JDK 11 (and before on Windows) where a channel is only closed after it is cancelled and flushed from its selector.

Tests in akka-actor-tests are now fixed. Something is still broken in akka-stream.

Fixes #25733, #25913.

jrudolph committed:
=act #25733 close channels from SelectionHandler and force flushing from selector

This is needed because starting from JDK 11 all platforms will only finally
close a channel when it is flushed from a selector.

Fixes #25733.

jrudolph commented:
Seems the stream tests fail because they depend on 

> the invariant that closing of a channel happens-before terminating the associated TcpConnection actor because the closing code may happen only in postStop where we cannot synchronously block until the selector has finally completed the channel closing.

And I tried getting away with not providing that guarantee but seems we might need it.

johanandren commented:
Are you pursuing some other idea that will work for streams or should I still review this one?

jrudolph commented:
> Are you pursuing some other idea that will work for streams or should I still review this one?

It should be a smallish update to TcpConnection. But if you wait a while maybe you can review the whole.

jrudolph commented:
Didn't manage to finish it today. The stream tests are currently super-flaky for me and I haven't found out why.

jrudolph commented:
Split up fixes unrelated to the selector issue out into #26038. Hopefully both PRs validate separately on JDK8.

This one was rebased and cleaned up. Once PR validation has cleared it, it should be finally ready for review.

jrudolph commented on akka-actor-tests/src/test/scala/akka/io/TcpConnectionSpec.scala:
```diff
@@ -915,9 +923,11 @@ class TcpConnectionSpec extends AkkaSpec("""
         def enableInterest(op: Int): Unit = interestCallReceiver.ref ! op
         def disableInterest(op: Int): Unit = interestCallReceiver.ref ! -op
-        def cancel(): Unit = ()
+        def cancelAndClose(andThen: () ⇒ Unit): Unit = onCancelAndClose(andThen)
```
Yep, we don't use the real SelectorHandler impl here, so we need to replicated the correct behavior.

jrudolph commented on akka-actor-tests/src/test/scala/akka/io/TcpIntegrationSpec.scala:
```diff
@@ -152,6 +156,13 @@ class TcpIntegrationSpec extends AkkaSpec("""
       override def bindOptions = List(SO.SendBufferSize(1024))
       override def connectOptions = List(SO.ReceiveBufferSize(1024))
+
+      serverHandler.send(serverConnection, Close)
```
Test was missing cleanup for some reason.

jrudolph commented on akka-actor-tests/src/test/scala/akka/io/TcpListenerSpec.scala:
```diff
@@ -143,13 +145,18 @@ class TcpListenerSpec extends AkkaSpec("""
         def enableInterest(op: Int): Unit = interestCallReceiver.ref ! op
         def disableInterest(op: Int): Unit = interestCallReceiver.ref ! -op
-        def cancel(): Unit = ()
+        def cancelAndClose(andThen: () ⇒ Unit): Unit = {
```
Same here, we need to replicate actual implementation.

jrudolph commented on akka-actor/src/main/scala/akka/io/TcpConnection.scala:
```diff
@@ -70,9 +71,9 @@ private[io] abstract class TcpConnection(val tcp: TcpExt, val channel: SocketCha
-      doRead(info, None) // immediately try reading, pullMode is handled by readingSuspended
+      // if we are in push mode or already have resumed reading in pullMode while waiting for Register
+      // then register OP_READ interest
+      if (!pullMode || ( /*pullMode && */ !readingSuspended)) resumeReading(info)
```
Previously, we were calling `doRead` here directly, however, `doRead` may already try to close the connection and call `context.become(unregistering)` which would then be lost if `become` a few lines below.

It seems the `doRead` was an optimization in the first place trying to instantly read something without checking the selector first. Now, we always go through the selector, which is more consistent anyway.

jrudolph commented on akka-actor/src/main/scala/akka/io/TcpListener.scala:
```diff
@@ -98,10 +97,13 @@ private[io] class TcpListener(
       log.debug("Unbinding endpoint {}", localAddress)
-      channel.close()
-      // see https://github.com/akka/akka/issues/20282
-      if (Helpers.isWindows) registration.enableInterest(1)
```
No extra code paths any more for Windows.

jrudolph commented on akka-actor/src/main/scala/akka/io/TcpOutgoingConnection.scala:
```diff
@@ -53,6 +54,7 @@ private[io] class TcpOutgoingConnection(
 
   def receive: Receive = {
     case registration: ChannelRegistration ⇒
+      setRegistration(registration)
```
Necessary to keep our registration here early, so we can actually close the connection on error.

patriknw reviewed (approved):
LGTM, as far as I can see

jrudolph commented:
Running JDK 11 nightly against both this and #26038 in https://jenkins.akka.io:8498/job/akka-nightly-jdk11/80/console.

johanandren commented on akka-actor/src/main/scala/akka/io/SelectionHandler.scala:
```diff
@@ -208,17 +210,38 @@ private[io] object SelectionHandler {
+        new Task {
+          def tryRun(): Unit = {
+            Try(key.cancel())
+            Try(key.channel().close())
```
Is it ok to do the same on Java 8 and 11 with this fix?

jrudolph replied:
Yes, that was the goal. We do a bit more than before on JDK 8. I guess, it's a good change anyway because it might fix https://github.com/akka/akka/issues/25605 as well.

johanandren reviewed (approved):
Tests green! LGMT. Fingers crossed for the nightly/repeated tests.
